beets/dbcore/db.py

```diff
@@ -1082,6 +1051,43 @@ def _make_attribute_table(self, flex_table):
                 """.format(flex_table))
 
     # Querying.
+    @staticmethod
+    def print_query(sql, subvals):
+        """If debugging, replace placeholders and print the query."""
+        if DEBUG:
+            topr = sql
+            for val in subvals:
+                topr = topr.replace("?", str(val), 1)
+            print(topr)
+
+    def _get_matching_ids(self, model, where, subvals):
+        """Return ids of entities which match the given filter (`where` clause).
+        This function is called only if we filter by at least one flexible
+        attribute field.
+
+        Since we cannot tell which entity the flexible attributes belong to
+        (or even whether they exist), we must join the related entity and query
+        each flexible attribute tables separately.
+
+        Since these queries only return IDs of entities _matching_ the filter,
+        they are quick regardless of whether the field is a model or a flexible
+        attribute, and how big is user's music library.
+        """
+        table = model._table
+        id_field = f"{table}.id"
+        join_tmpl = "LEFT JOIN {} ON {} = entity_id"
+        joins = [join_tmpl.format(model._flex_table, id_field)]
+        if table == "items":
+            joins.append(join_tmpl.format("album_attributes", "items.album_id"))
```

Comment:
[longer-term comment] I know this is a tall order, but it would be nice someday if `dbcore` didn't need to know anything about beets models. (It's meant to be a mostly-generic database abstraction layer.) This may be unavoidable, at least in some form, for dealing with the "item-to-album fallback." But I just wanted to put that lofty goal out here.

Reply:
Hehe, I like your thinking! @wisp3rwind made the same note here: https://github.com/beetbox/beets/pull/4362#discussion_r895240311 :)

Luckily in this design that's an easy adjustment, see 56909464